dial_queue.go

```diff
@@ -0,0 +1,269 @@
+// - we scale down when we've been idle for a while waiting for new dial attempts.
+// - we scale down when we complete a dial and realise nobody was waiting for it.
+//
+// Dialler throttling (e.g. FD limit exceeded) is a concern, as we can easily spin up more workers to compensate, and
+// end up adding fuel to the fire. Since we have no deterministic way to detect this for now, we hard-limit concurrency
+// to DialQueueMaxParallelism.
+func newDialQueue(ctx context.Context, target string, in *queue.ChanQueue, dialFn func(context.Context, peer.ID) error, nConsumers int) *dialQueue {
+	sq := &dialQueue{
+		ctx:           ctx,
+		dialFn:        dialFn,
+		nWorkers:      DialQueueMinParallelism,
+		scalingFactor: 1.5,
+
+		in:  in,
+		out: queue.NewChanQueue(ctx, queue.NewXORDistancePQ(target)),
+
+		growCh:    make(chan struct{}, nConsumers),
+		shrinkCh:  make(chan struct{}, 1),
+		waitingCh: make(chan waitingCh, nConsumers),
+		dieCh:     make(chan struct{}, DialQueueMaxParallelism),
+	}
+	for i := 0; i < DialQueueMinParallelism; i++ {
+		go sq.worker()
+	}
+	go sq.control()
+	return sq
+}
+
+func (dq *dialQueue) control() {
+	var (
+		p              peer.ID
+		dialled        = dq.out.DeqChan
+		resp           waitingCh
+		waiting        <-chan waitingCh
+		lastScalingEvt = time.Now()
+	)
+	for {
+		// First process any backlog of dial jobs and waiters -- making progress is the priority.
+		// This block is copied below; couldn't find a more concise way of doing this.
+		select {
+		case <-dq.ctx.Done():
+			return
+		case p = <-dialled:
+			dialled, waiting = nil, dq.waitingCh
+			continue // onto the top.
+		case resp = <-waiting:
+			// got a channel that's waiting for a peer.
+			log.Debugf("delivering dialled peer to DHT; took %dms.", time.Now().Sub(resp.ts)/time.Millisecond)
+			resp.ch <- p
+			close(resp.ch)
+			dialled, waiting = dq.out.DeqChan, nil // stop consuming waiting jobs until we've cleared a peer.
+			continue                               // onto the top.
+		default:
+			// there's nothing to process, so proceed onto the main select block.
+		}
+
+		select {
+		case <-dq.ctx.Done():
+			return
+		case p = <-dialled:
+			dialled, waiting = nil, dq.waitingCh
+		case resp = <-waiting:
+			// got a channel that's waiting for a peer.
+			log.Debugf("delivering dialled peer to DHT; took %dms.", time.Now().Sub(resp.ts)/time.Millisecond)
+			resp.ch <- p
+			close(resp.ch)
+			dialled, waiting = dq.out.DeqChan, nil // stop consuming waiting jobs until we've cleared a peer.
+		case <-dq.growCh:
+			if time.Now().Sub(lastScalingEvt) < DialQueueScalingMutePeriod {
+				continue
+			}
+			dq.grow()
+			lastScalingEvt = time.Now()
+		case <-dq.shrinkCh:
+			if time.Now().Sub(lastScalingEvt) < DialQueueScalingMutePeriod {
+				continue
+			}
+			dq.shrink()
+			lastScalingEvt = time.Now()
+		}
+	}
+}
+
+func (dq *dialQueue) Consume() (<-chan peer.ID, error) {
```

Comment:
Instead of returning an error, maybe return a closed channel to indicate that we're done? It's unclear what do do with the error because it *could* mean "we're done" or it could mean "you did something wrong". Even if we get rid of the second case (panic), this still doesn't really feel like an "error" case.

Reply:
I think that's fair.